### PR TITLE
MEN-5219: Implement integration with /etc/grub.d framework.

### PR DESCRIPTION
**Pull request by kacf**
This means that running `grub-install` and `update-grub` on the running
device does not ruin the Mender integration, but results in newly
generated scripts with Mender features included.

For details about the implementation I refer to the two README files
inside the `grub.d` folder. I'm not going to repeat it here. However, I
want to highlight a few key changes that were made while implementing
this:

* Since the scripts now share a location with upstream GRUB scripts, as
  well as possible custom modifications, rename all grub-mender-grubenv
  script files to have a "mender_" prefix after the index, so their
  purpose is clear.

* Since we are now using the upstream version of the GRUB binary we need
  to adapt to the capabilities it has. The `hashsum` command is not
  included in that binary, so refactor the environment handling code to
  not use it. This is pretty easy, since the file we were checksumming
  before only contains one variable, and its content can only be 0 or 1,
  so we can just verify its one of those.

* Made a new pass looking at all the places we use `load_env` and made
  sure that `--skip-sig` is used correctly. This does not matter for
  GRUB binaries shipped by Ubuntu or Debian, since they don't enforce
  signatures, but it does matter if someone builds their own enforcing
  binary.

* Makefile targets have changed. The new default for `install` is to
  install the grub.d integration, and the standalone variant is now
  available by calling make with `install-standalone` instead. There are
  several new sub-targets as well to deal with various specialized
  circumstances. I refer to the file itself for details.

* The default location for the environment has moved from
  `/boot/efi/EFI/BOOT/mender_grubenv*` to
  `/boot/efi/grub-mender-grubenv/mender_grubenv*`. This was done
  primarily because `grub-install` puts its own hardcoded `grub.cfg` in
  the former folder. Our custom boot scripts therefore live in the
  latter folder, so the environment should live there as well. See also
  the next point.

* `/etc/mender_grubenv.config` is now optional and is no longer
  installed by default. This makes it possible for the tool to
  auto-detect where the environment is, so that it can keep supporting
  platforms that still have it in the old location. The configuration
  file is still supported though, for those who have special needs.

Changelog: Title

Signed-off-by: Kristian Amlie <kristian.amlie@northern.tech>

### CODE REVIEW

**Comment by kacf** [comment]
I recommend reading the READMEs first for an overview.

File: grub.d/README-internal.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+`grub.d` integration
```
Comment (TheMeaningfulEngineer) [inline grub.d/README-internal.md:1]: @kacf 
I'm finding it hard to follow the thread of "Two styles".
I assume you're naming them `grub-mender-grubenv` and `grub.d`. 
Would appreciate a higher level introduction on why they are here and what they do. Here is an attempt:

> During boot time, grub needs the scripts in this directory to exist in a certain location on the boot partition.
> This process of script installation also involves modifying the scripts (concatenating/renaming). 
> 
> We have two mechanisms for this:
> 
> * `grub-mender-grubenv`
>     * how we historically did it
>     * no external dependencies on grub tools
>     * breaks the bootlader if groob tools are used in paralel
> * `grub.d` - 
>     * standard way used used in Debian
>     * uses `update-grub` and `grub-mkconfig`


Is this correctly interpreted?

File: README.md [hunk 2]
```diff
@@ -29,14 +29,39 @@ make install
...
+To install this variant, run this:
+
+```
+make
+make install-standalone
+```
+
+You can also control the location of the installed environment files on the
+final target with `ENV_DIR`. The default is `/boot/efi/EFI/BOOT` which assumes
+that the [EFI
```
Comment (TheMeaningfulEngineer) [inline README.md:59]: The main README should somehow reference the internal docs also (ones in `grub-mender-grubenv/grub.d/`).
This would allow for a smoother explanation on what those documents describe.

File: README.md [hunk 3]
```diff
@@ -29,14 +29,39 @@ make install
...
+`/etc/grub.d` scripts, which means you need to run `grub-install` and
+`update-grub` after installation on the platform. This normally has to be done
+while running the device, but you can use
+[mender-convert](https://github.com/mendersoftware/mender-convert) as a way to
+do this at build time.
+
+If you want to install only the tools, or only the script and environment files,
+run `make install-tools` or `make install-grub.d-boot-files`, respectively.
+
+### Standalone boot scripts
```
Comment (TheMeaningfulEngineer) [inline README.md:42]: I'm assuming we want to move away from this as a general rule,
in that case it should be presented in a way that it's not a first class citizen but instead a "custom do only if you know why you're doing it" case. 

Also is the following mapping correct?:
* [Installing](https://github.com/kacf/grub-mender-grubenv/tree/grub.d_integration#installing) = `grub.d flow`
* Standalone boot scripts = `grub-mender-grubenv flow`

File: README.md [hunk 4]
```diff
@@ -29,14 +29,39 @@ make install
...
-files on the final target with `ENV_DIR`. The default is `/boot/efi/EFI/BOOT`
-which assumes that the [EFI
+`install` target. By default, the installation is integrated with the
+`/etc/grub.d` scripts, which means you need to run `grub-install` and
+`update-grub` after installation on the platform. This normally has to be done
+while running the device, but you can use
+[mender-convert](https://github.com/mendersoftware/mender-convert) as a way to
+do this at build time.
+
+If you want to install only the tools, or only the script and environment files,
```
Comment (TheMeaningfulEngineer) [inline README.md:39]: What are the "tools which one installs"? `grub-install` and those?

File: README.md (hunk 2 above)
Comment (kacf) [reply README.md:59]: Good thinking, I'll add a pointer here.

File: README.md (hunk 3 above)
Comment (kacf) [reply README.md:42]: It should be the default, but we can not move away from it entirely, because it is still the only way to do it in Yocto, and this is unlikely to change. But still a good point about a warning, I'll add something like that, and perhaps mention that Yocto is a place where it's used also.

The mapping is correct.

File: README.md (hunk 4 above)
Comment (kacf) [reply README.md:39]: Ah no, it's `grub-mender-grubenv-set` and `grub-mender-grubenv-print`. I should probably make that clear!

File: grub.d/90_mender_generate_dual_rootfs_grub [hunk 5]
```diff
@@ -0,0 +1,37 @@
+set -e
+
+if [ "$GRUB_MENDER_GRUBENV_CFG_GENERATION" != "true" ]; then
+    if [ "$(${grub_probe} --target device /)" = "$(${grub_probe} --target device /boot/grub)" ]; then
+        cat 1>&2 <<'EOF'
+ERROR: /boot/grub is on the root filesystem. This is not supported when using
+Mender with dual rootfs setup.
+Please make sure that it is either a symlink to a boot partition (usually it
+points to `efi/grub-mender-grubenv`), or that a boot partition is mounted there.
+EOF
```
Comment (lluiscampos) [inline grub.d/90_mender_generate_dual_rootfs_grub:10]: Shouldn't we `exit` here?

File: grub.d/default/mender-os-probe-skip.cfg [hunk 6]
```diff
@@ -0,0 +1,7 @@
+. /usr/share/grub/grub-mkconfig_lib
+
+for device in ${mender_kernel_root_base}${mender_rootfsa_part} ${mender_kernel_root_base}${mender_rootfsb_part}; do
```
Comment (merlin-northern) [inline grub.d/default/mender-os-probe-skip.cfg:3]: do we need double quotes in here?

File: grub.d/default/xx_mender_font_and_theme_handling.cfg [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	for basename in unicode unifont ascii; do
+	    path="${dir}/${basename}.pf2"
+	    if [ -e "${path}" ]; then
+		GRUB_FONT="${path}"
+	    else
+		continue
+	    fi
+	    break 2
+	done
+    done
```
Comment (lluiscampos) [inline grub.d/default/xx_mender_font_and_theme_handling.cfg:13]: Could you fix the indention here?

**Review by lluiscampos — APPROVED** [review]
Obviously, some details went over my head... I also didn't review the tests. But overall the work looks good, and I don't need much more evidence that the demo you did yesterday :)

I dropped few comments but pre-aproving as I trust your judgment in such things.

File: tests/test_makefile.sh [hunk 8]
```diff
@@ -10,43 +10,228 @@ function cleanup() {
...
+boot/efi/grub-mender-grubenv/mender_grubenv1/lock.sha256sum
+boot/efi/grub-mender-grubenv/mender_grubenv2/env
+boot/efi/grub-mender-grubenv/mender_grubenv2/lock
+boot/efi/grub-mender-grubenv/mender_grubenv2/lock.sha256sum
+EOF
+diff -u "$TMPDIR/expected.log" "$TMPDIR/actual.log"
+
+make_check_clean_target uninstall-standalone-boot-files
+
+rm -rf "$TMPDIR"/*
```
Comment (merlin-northern) [inline tests/test_makefile.sh:41]: despite this being a test, and TMPDIR not being empty most probably, I would consider:
```bash
[ "${TMPDIR}" != "" ] && rm -rf "$TMPDIR"/*
```
I know, paranoia mode:on.

**Review by merlin-northern — APPROVED** [review]
I left some comments, but I salute you this is great work -- I do not understand half of it. treat my comments as suggestions.

File: grub.d/90_mender_generate_dual_rootfs_grub (hunk 5 above)
Comment (kacf) [reply grub.d/90_mender_generate_dual_rootfs_grub:10]: Yes! Good catch.

File: 90_mender_boot_grub.cfg [hunk 9]
```diff
@@ -1,34 +1,7 @@
...
-    elif [ "${mender_boot_part}" = "${mender_rootfsb_part}" ]; then
-        mender_root="PARTUUID=${mender_rootfsb_uuid}"
-    fi
-else
-    mender_root="${mender_kernel_root_base}${mender_boot_part}"
-fi
-
 if [ "${drop_to_grub_prompt}" = "no" ]; then
-    if linux ${mender_kernel_path}/${kernel_imagetype} root=${mender_root} ${bootargs}; then
+    if linux ${mender_kernel_path}/${kernel_imagetype} root=${mender_kernel_root} ${bootargs}; then
```
[the remark this answers is not in the document]
Comment (kacf) [reply 90_mender_boot_grub.cfg:4]: This is pre-existing, but yes, it is better. I'll change this and the other one below.

File: grub-mender-grubenv-print [hunk 10]
```diff
@@ -2,34 +2,44 @@
...
+
+# In the past, we also used a sha256sum checksum file. The boot script doesn't
+# use this anymore because the `hashsum` command is not included in all distro
+# editions of GRUB. But because the sha256sum checksum also makes sure the
+# format of the whole environment file is valid, not just the variable, we have
+# kept using it in this command line utility for better detection of corruption.
 
 if [ -r /etc/mender_grubenv.config ]; then
     ENV_DIR="$(egrep '^ENV_DIR *= *' /etc/mender_grubenv.config | sed -e 's/^ENV_DIR *= *//' | tail -n 1)"
 fi
```
[the remark this answers is not in the document]
Comment (kacf) [reply grub-mender-grubenv-print:39]: Yes, the file is still supported for those who have special needs. But we won't install it by default anymore.

File: grub.d/00_90_mender_boot_selected_rootfs [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    reboot
+elif [ "${check_signatures}" == "enforce" ]; then
+    echo "Returned from rootfs boot script. This could mean that the currently selected rootfs is corrupt. But there is no upgrade in progress, and Signature Verification is on, so there is nothing we can do. Shutting down."
+    sleep 10
+    halt
+else
+    echo "Returned from rootfs boot script. This could mean that the currently selected rootfs is corrupt. But there is no upgrade in progress, so dropping to prompt."
+    sleep 10
+    normal_exit
+fi
```
[the remark this answers is not in the document]
Comment (kacf) [reply grub.d/00_90_mender_boot_selected_rootfs:18]: 👍

File: grub.d/90_mender_generate_dual_rootfs_grub [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    # so that we don't print them a second time.
+    GRUB_MENDER_GRUBENV_CFG_GENERATION=true grub-mkconfig -o $tempfile 2>/dev/null
+
+    # The first boot entry, which is for the partition which generates this boot
+    # script, and the rootfs used by Mender, should not have "root=" statement,
+    # since this is set by Mender, and should not be set by the boot entry
+    # itself. This guarantees that the rootfs can be loaded from either of the
+    # two dual rootfs partitions.
+    if egrep -m 1 -B 3 '^\s+([Ll]oading [Ll]inux|linux\s)' $tempfile | egrep '^\s*(set\s+)?root=' 1>&2; then
+        cat 1>&2 <<EOF
```
[the remark this answers is not in the document]
Comment (kacf) [reply grub.d/90_mender_generate_dual_rootfs_grub:26]: This is intentional, they have different meanings. `EOF` provides variable and command expansion inside the document, whereas `'EOF'` will be rendered verbatim.

File: grub.d/default/extra/mender-offline-grub-probe-helper.cfg [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+# time image using mender-convert. GRUB provides a way to swap out its
+# `grub-probe` tool, which is what this script uses to replace live probes with
+# preconfigured values.
+
+# The offline mode is designed to run in a container or in chroot, never outside.
+if type systemd-detect-virt 2>&1 >/dev/null && \
+    ! systemd-detect-virt --container >/dev/null && \
+    ! systemd-detect-virt --chroot >/dev/null; then
+
+    echo "The /etc/default/grub.d/mender-offline-grub-probe-helper.cfg file is only meant to run in a container or chroot as part of a build process. Please remove it before running on a real device." 1>&2
```
[the remark this answers is not in the document]
Comment (kacf) [reply grub.d/default/extra/mender-offline-grub-probe-helper.cfg:12]: No, because the installed path is not the same as the source path. It's in the `extra` source folder because it is normally only installed when using mender-convert, and even then, only temporarily. So if it is still here when running on a real device, someone made a mistake, and it should be deleted from the mentioned path.

File: tests/test_makefile.sh (hunk 8 above)
Comment (kacf) [reply tests/test_makefile.sh:41]: Hehe, probably a good point. I'll do it in a simpler way and simply exit early if it fails to be set at the top.

File: grub.d/default/extra/mender-offline-grub-probe-helper.cfg [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+# `update-grub` in a container or chroot. This is necessary when making a build
+# time image using mender-convert. GRUB provides a way to swap out its
+# `grub-probe` tool, which is what this script uses to replace live probes with
+# preconfigured values.
+
+# The offline mode is designed to run in a container or in chroot, never outside.
+if type systemd-detect-virt 2>&1 >/dev/null && \
+    ! systemd-detect-virt --container >/dev/null && \
+    ! systemd-detect-virt --chroot >/dev/null; then
+
```
[the remark this answers is not in the document]
Comment (kacf) [reply grub.d/default/extra/mender-offline-grub-probe-helper.cfg:11]: I think it helps separate the condition from the action. I'm not completely sure what is our preferred style with regards to indentation here.

**Comment by kacf** [comment]
Updated! All things not commented on have been fixed.

**Review by merlin-northern — APPROVED** [review]

**Review by oleorhagen — APPROVED** [review]
Oh my... Took me a while to get through!

Impressive work!

File: grub.d/default/xx_mender_font_and_theme_handling.cfg [hunk 15]
```diff
@@ -1,3 +1,7 @@
+pkgdatadir="${pkgdatadir:-/usr/share/grub}"
+
+. "${pkgdatadir}/grub-mkconfig_lib"
```
Comment (merlin-northern) [inline grub.d/default/xx_mender_font_and_theme_handling.cfg:3]: this file always exists, right?

File: 04_mender_setup_env_grub_functions.cfg [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+        load_env --skip-sig --file ${MENDER_ENV1} bootcount mender_boot_part upgrade_available mender_systemd_machine_id
+    fi
+
+    if ! mender_check_grubenv_valid; then
+        if [ "${check_signatures}" == "enforce" ]; then
+            echo "Unverified environment and signatures enabled. Rebooting."
+            sleep 10
+            reboot
+        fi
+    fi
```
[the remark this answers is not in the document]
Comment (kacf) [reply 04_mender_setup_env_grub_functions.cfg:106]: Good point, there should be an error path here, actually. At this point there is little we can do, but I reckon that setting `mender_boot_part=${mender_rootfsa_part}` gives at least a chance that the system might boot, even if the partition may be wrong.

I don't think it can be used for downgrade attacks, because you'd need to have root access to corrupt the environment in this way, and if you had that you could just change the boot script directly.

File: 95_mender_try_to_recover_grub.cfg [hunk 17]
```diff
@@ -1,4 +1,4 @@
-if [ "${drop_to_grub_prompt}" = "no" ]; then
+if [ "${drop_to_grub_prompt}" != "yes" ]; then
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply 95_mender_try_to_recover_grub.cfg:1]: :older_man:

File: grub-mender-grubenv-print [hunk 18]
```diff
@@ -2,34 +2,44 @@
...
+if [ -z "$ENV_DIR" -a -d /boot/efi/grub-mender-grubenv ]; then
```
[the remark this answers is not in the document]
Comment (kacf) [reply grub-mender-grubenv-print:40]: Hmmmm, not sure I understand..? The check below is only true if this *doesn't* execute, because we set `ENV_DIR` in the `if `body.

[the remark this answers is not in the document]
Comment (kacf) [reply grub-mender-grubenv-print:40]: I suppose a better way to express it is:

```bash
if [ -z "$ENV_DIR" ]; then
    if [ -d /boot/efi/grub-mender-grubenv ]; then
        ENV_DIR=/boot/efi/grub-mender-grubenv
    else
        ENV_DIR=/boot/efi/EFI/BOOT
    fi
fi
```

Maybe I'll switch to that, I think it's clearer.

File: grub.d/08_mender_rollback.cfg [hunk 19]
```diff
@@ -0,0 +1,14 @@
+if [ "${upgrade_available}" = "1" ]; then
+    menuentry "Roll back latest upgrade" --id mender_rollback {
+        reboot
+    }
+
+    # Ideally we would have liked to use a string in this "fallback" variable,
+    # but because there is a bug in GRUB which accepts only indexes for
+    # "fallback", we need to put the menuentry here in the first position, and
+    # then use index 0 in "fallback". In a later script we will set "default",
+    # which does accept strings, so that we don't boot the fallback by default.
```
[the remark this answers is not in the document]
Comment (kacf) [reply grub.d/08_mender_rollback.cfg:10]: Exactly, the ID does not work.

The GRUB docs [explicitly say](https://salsa.debian.org/grub-team/grub/-/blob/master/util/grub.d/README) that boot entries start at 10, so 08 should be fine. We don't want it too early, because there should be room for adding initialization scripts before it. In the end the choice of 08 is somewhat arbitrary of course, but seems like a reasonable compromise.

File: grub-mender-grubenv-print (hunk 18 above)
[the remark this answers is not in the document]
Comment (oleorhagen) [reply grub-mender-grubenv-print:40]: Ahh, yes. Makes sense. And much better imo :)

File: grub.d/90_mender_generate_dual_rootfs_grub [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+    # from grub-mkconfig should already have been printed. Therefore hide stderr
+    # so that we don't print them a second time.
+    GRUB_MENDER_GRUBENV_CFG_GENERATION=true grub-mkconfig -o $tempfile 2>/dev/null
+
+    # The first boot entry, which is for the partition which generates this boot
+    # script, and the rootfs used by Mender, should not have "root=" statement,
+    # since this is set by Mender, and should not be set by the boot entry
+    # itself. This guarantees that the rootfs can be loaded from either of the
+    # two dual rootfs partitions.
+    if egrep -m 1 -B 3 '^\s+([Ll]oading [Ll]inux|linux\s)' $tempfile | egrep '^\s*(set\s+)?root=' 1>&2; then
```
[the remark this answers is not in the document]
Comment (kacf) [reply grub.d/90_mender_generate_dual_rootfs_grub:26]: Hehe, alright, alright, you convinced me! These are not among the most common ones, so I switched!

File: grub.d/Makefile [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+
+install:
+	install -m 755 -d $(DESTDIR)$(sysconfdir)/grub.d
+	set -e; for file in $(srcdir)/[0-9][0-9]_*; do \
+		echo "Installing $$file"; \
+		case $$file in \
+		*.cfg) \
+			dest=$(DESTDIR)$(sysconfdir)/grub.d/$$(basename $$file); \
+			dest=$${dest%.cfg}; \
+			echo "cat <<'END_OF_MENDER_GRUBENV_CFG_FILE'" > $$dest; \
```
[the remark this answers is not in the document]
Comment (kacf) [reply grub.d/Makefile]: If you like this, check out [this beauty](https://github.com/cfengine/core/blob/master/tests/acceptance/testall#L1273-L1301) from CFengine. I'm sure they're still cursing me for it!

File: 04_mender_setup_env_grub_functions.cfg [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+# Note that Secure Boot and GRUB signatures are two different things, and here
+# we are talking about the latter.
+
+MENDER_ENV1=${prefix}/mender_grubenv1/env
+MENDER_LOCK1=${prefix}/mender_grubenv1/lock
+MENDER_ENV2=${prefix}/mender_grubenv2/env
+MENDER_LOCK2=${prefix}/mender_grubenv2/lock
+
+function mender_check_and_restore_env {
+    editing=invalid
```
[the remark this answers is not in the document]
Comment (kacf) [reply 04_mender_setup_env_grub_functions.cfg:20]: Yes, strictly speaking any nonzero value marks a corrupt environment, so setting it to "invalid" is at best just a way to figure out where the "invalidness" is coming from.

File: grub.d/default/xx_mender_font_and_theme_handling.cfg (hunk 15 above)
Comment (kacf) [reply grub.d/default/xx_mender_font_and_theme_handling.cfg:3]: Yes, it is a standard file included by several of the upstream scripts as well.

File: grub.d/Makefile (hunk 21 above)
[the remark this answers is not in the document]
Comment (oleorhagen) [reply grub.d/Makefile]: hahah, that is a real beaut!

**Comment by kacf** [comment]
Alright, fixed and updated.

**Comment by oleorhagen** [comment]
:+1:
